File: wigglystuff/observable_plot.py
"""ObservablePlot widget for running Observable Plot code in the notebook."""

from pathlib import Path
from typing import Any, Dict, Optional

import anywidget
import traitlets


class ObservablePlot(anywidget.AnyWidget):
    """Run `Observable Plot <https://observablehq.com/plot/>`_ code in a notebook.

    Observable Plot is a JavaScript charting library. This widget is a thin
    runner: it loads Plot (and d3) from a CDN, then evaluates a blob of
    JavaScript you supply (the ``code`` traitlet) exactly as you would write it
    in an Observable notebook — a bare ``Plot.plot({...})`` expression that
    returns a DOM node. The returned node is mounted inline; Python's job is to
    hand over the source, the data, and the sizing.

    Python values are injected into the code's scope by name via ``variables``:
    each key becomes an in-scope JavaScript variable holding the (JSON-ified)
    value. So ``variables={"vacancies": df}`` lets your code reference
    ``vacancies`` directly::

        Plot.plot({
            marks: [
                Plot.barY(vacancies, { x: "month", y: "vacancies" }),
                Plot.ruleY([0]),
            ],
        })

    ``Plot`` and ``d3`` are always in scope, along with ``container`` (the
    target DOM element), the ``width`` / ``height`` ints, and the widget
    ``model``. Pandas / polars DataFrames and numpy arrays in ``variables`` are
    converted to plain JSON structures (records / lists) before syncing.

    The code is evaluated verbatim in the browser, so only pass JavaScript you
    trust (typically your own).

    Reassigning ``code``, ``variables``, ``width``, ``height``, or ``version``
    re-runs the code and rebuilds the chart. Observable Plot has no incremental
    redraw — each run produces a fresh SVG — but the new node is swapped in
    atomically (the previous chart stays put until it is ready), so updates
    driven by e.g. a slider do not flash a blank frame.

    Examples:
        ```python
        import marimo as mo
        import pandas as pd
        from wigglystuff import ObservablePlot

        df = pd.DataFrame({"month": ["Jan", "Feb", "Mar"], "vacancies": [3, 5, 2]})
        code = '''
            Plot.plot({
                marks: [
                    Plot.barY(vacancies, { x: "month", y: "vacancies" }),
                    Plot.ruleY([0]),
                ],
            })
        '''
        mo.ui.anywidget(ObservablePlot(code, variables={"vacancies": df}))
        ```

        Point at a local file or a URL instead of an inline string — the first
        argument accepts any of the three forms:

        ```python
        ObservablePlot("charts/bars.js", variables={"vacancies": df})
        ObservablePlot("https://example.com/chart.js")
        ```
    """

    _esm = Path(__file__).parent / "static" / "observable-plot.js"
    _css = Path(__file__).parent / "static" / "observable-plot.css"

    code = traitlets.Unicode("").tag(sync=True)
    variables = traitlets.Dict({}).tag(sync=True)
    width = traitlets.Int(500).tag(sync=True)
    height = traitlets.Int(300).tag(sync=True)
    version = traitlets.Unicode("latest").tag(sync=True)
    error = traitlets.Unicode("").tag(sync=True)

    def __init__(
        self,
        code: Optional[str] = None,
        *,
        src: Optional[str] = None,
        variables: Optional[Dict[str, Any]] = None,
        width: int = 500,
        height: int = 300,
        version: str = "latest",
        **kwargs: Any,
    ) -> None:
        """Create an ObservablePlot widget.

        The plot source is resolved in Python, so the browser always receives
        plain JavaScript. ``code`` (or ``src``) may be any of:

        * an inline JavaScript string,
        * a path to a local ``.js`` file, or
        * an ``http(s)://`` URL.

        URLs and file paths are fetched/read here at construction time; the
        detection is by value, so ``ObservablePlot("https://.../chart.js")`` and
        ``ObservablePlot("Plot.plot({...})")`` both do the right thing.

        Args:
            code: Inline JavaScript, a local file path, or an ``http(s)://``
                URL. Mutually exclusive with ``src``.
            src: An explicit alias for ``code`` (same accepted forms), handy
                when the intent is "load from here". Mutually exclusive with
                ``code``.
            variables: Mapping of name -> value injected into the code's scope
                as JavaScript variables. DataFrames and numpy arrays are
                converted to JSON-serializable structures.
            width: Container width in pixels.
            height: Container height in pixels.
            version: The `@observablehq/plot` version to load from the CDN.
                Defaults to ``"latest"``; pin an exact version (e.g. ``"0.6.17"``)
                for reproducibility.
            **kwargs: Forwarded to ``anywidget.AnyWidget``.

        Raises:
            ValueError: If neither or both of ``code`` and ``src`` are given.
        """
        if (code is None) == (src is None):
            raise ValueError("Provide exactly one of `code` or `src`.")

        source = code if code is not None else src
        super().__init__(
            code=self._resolve_source(source),
            variables=variables if variables is not None else {},
            width=width,
            height=height,
            version=version,
            **kwargs,
        )

    @traitlets.validate("variables")
    def _validate_variables(self, proposal: Any) -> Dict[str, Any]:
        # Convert on every assignment, so `widget.variables = {"df": df}` works
        # the same as passing it to the constructor.
        return self._convert_variables(proposal["value"])

    @staticmethod
    def _resolve_source(source: str) -> str:
        """Return plot JS: fetch a URL, read a local file, or pass JS through."""
        text = str(source)
        if text.startswith("http://") or text.startswith("https://"):
            import urllib.request

            with urllib.request.urlopen(text) as response:
                return response.read().decode("utf-8")
        # A path to an existing file? Read it. Otherwise treat it as inline JS.
        # (Multi-line JS makes an invalid path, so is_file() is safely False.)
        try:
            path = Path(text)
            if path.is_file():
                return path.read_text(encoding="utf-8")
        except (OSError, ValueError):
            pass
        return text

    @classmethod
    def _convert_variables(
        cls, variables: Optional[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """Convert each injected variable to a JSON-serializable form."""
        if variables is None:
            return {}
        if not isinstance(variables, dict):
            raise TypeError("`variables` must be a dict of name -> value.")
        return {str(name): cls._to_jsonable(value) for name, value in variables.items()}

    @staticmethod
    def _to_jsonable(value: Any) -> Any:
        """Coerce DataFrames / numpy types to plain JSON-serializable values.

        Duck-typed so pandas, polars, and numpy stay optional dependencies.
        Anything already JSON-native (lists, dicts, str, numbers, bool, None)
        passes straight through.
        """
        # polars DataFrame (check first -- polars also has .to_dict)
        if hasattr(value, "to_dicts") and callable(value.to_dicts):
            return value.to_dicts()
        # pandas DataFrame
        if hasattr(value, "to_dict") and callable(value.to_dict):
            return value.to_dict("records")
        # numpy array (or anything array-like exposing tolist)
        if hasattr(value, "tolist") and callable(value.tolist):
            return value.tolist()
        # numpy scalar -> Python native
        if hasattr(value, "item") and callable(value.item):
            return value.item()
        return value

File: wigglystuff/static/observable-plot.js
/**
 * ObservablePlot – run Observable Plot code in the notebook.
 *
 * Loads Observable Plot (and d3) from a CDN on first use, then evaluates the
 * user-supplied JS (the `code` traitlet) the way an Observable cell would: as
 * an expression that returns a DOM node (e.g. `Plot.plot({...})`). The returned
 * node is mounted into the container.
 *
 * `Plot`, `d3`, `container`, `width`, `height`, and `model` are always in
 * scope. Each key of the `variables` traitlet is injected as an additional
 * in-scope variable, so Python-supplied data can be referenced by name:
 *
 *     Plot.plot({
 *       marks: [ Plot.barY(vacancies, { x: "month", y: "vacancies" }) ],
 *     })
 *
 * See: https://observablehq.com/plot/
 */

// Cache the CDN import promise per Plot version so the library is fetched once.
// d3 is pinned to a single major line and shared across versions.
const _plotCache = {};
let _d3 = null;

function loadPlot(version) {
  if (!_plotCache[version]) {
    _plotCache[version] = import(
      "https://cdn.jsdelivr.net/npm/@observablehq/plot@" + version + "/+esm"
    );
  }
  return _plotCache[version];
}

function loadD3() {
  if (!_d3) {
    _d3 = import("https://cdn.jsdelivr.net/npm/d3@7/+esm");
  }
  return _d3;
}

// Build the runnable function for `code`. We first try the expression form
// (Observable cell semantics: the code IS the value to render). If that is a
// syntax error — e.g. the user wrote setup statements — we fall back to a block
// form where the user is expected to `return` the node themselves.
function compile(code, names) {
  const args = ["Plot", "d3", "container", "width", "height", "model", ...names];
  try {
    return new Function(...args, "return (async () => (\n" + code + "\n))();");
  } catch (err) {
    if (err instanceof SyntaxError) {
      return new Function(...args, "return (async () => {\n" + code + "\n})();");
    }
    throw err;
  }
}

function render({ model, el }) {
  el.classList.add("observable-plot-root");

  const container = document.createElement("div");
  container.className = "observable-plot-container";
  el.appendChild(container);

  const errorBox = document.createElement("pre");
  errorBox.className = "observable-plot-error";
  errorBox.style.display = "none";
  el.appendChild(errorBox);

  // Bumped on every (re)run so a stale async run can't clobber a newer one.
  let runToken = 0;

  function applySize() {
    container.style.width = model.get("width") + "px";
  }

  function showError(message) {
    errorBox.textContent = message;
    errorBox.style.display = "block";
    model.set("error", message);
    model.save_changes();
  }

  function clearError() {
    errorBox.textContent = "";
    errorBox.style.display = "none";
    if (model.get("error")) {
      model.set("error", "");
      model.save_changes();
    }
  }

  async function run() {
    const token = ++runToken;
    clearError();

    // `code` is already plain JavaScript — Python resolved any file/URL source.
    const code = model.get("code");
    if (!code) {
      container.replaceChildren();
      return;
    }

    let Plot, d3;
    try {
      [Plot, d3] = await Promise.all([loadPlot(model.get("version")), loadD3()]);
    } catch (err) {
      if (token !== runToken) return;
      container.replaceChildren();
      showError("Failed to load Observable Plot: " + (err && err.message ? err.message : err));
      return;
    }
    if (token !== runToken) return;

    const variables = model.get("variables") || {};
    const names = Object.keys(variables);

    // Observable Plot has no partial/incremental redraw — every run rebuilds the
    // whole SVG. To keep re-renders (e.g. a slider driving `variables`) from
    // flashing a blank frame, build into a detached `stage` and swap it into the
    // live container in a single atomic operation once the new node is ready.
    // The old chart stays visible until then.
    const stage = document.createElement("div");
    try {
      const fn = compile(code, names);
      const node = await fn(
        Plot,
        d3,
        stage,
        model.get("width"),
        model.get("height"),
        model,
        ...names.map((n) => variables[n])
      );
      if (token !== runToken) return;
      // Expression form returns the node; block form appends to `container`
      // (the stage) itself, in which case we mount the stage's contents.
      container.replaceChildren(node instanceof Node ? node : stage);
    } catch (err) {
      if (token !== runToken) return;
      container.replaceChildren();
      showError((err && err.stack) || String(err));
    }
  }

  applySize();
  run();

  model.on("change:code", run);
  model.on("change:variables", run);
  model.on("change:version", run);
  model.on("change:width", () => {
    applySize();
    run();
  });
  model.on("change:height", () => {
    applySize();
    run();
  });

  return () => {
    // Invalidate any in-flight run so it stops touching the removed DOM.
    runToken++;
  };
}

export default { render };
